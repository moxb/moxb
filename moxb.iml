<?xml version="1.0" encoding="UTF-8"?>
<module type="WEB_MODULE" version="4">
  <component name="NewModuleRootManager" inherit-compiler-output="true">
    <exclude-output />
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/packages/antd/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/meteor/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/moxb/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/packages/react-html/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/examples/src" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/packages/moxb/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/meteor/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/react-html/dist" />
      <excludeFolder url="file://$MODULE_DIR$/packages/antd/dist" />
      <excludeFolder url="file://$MODULE_DIR$/admin/node-installation" />
      <excludeFolder url="file://$MODULE_DIR$/admin/bin-tools" />
      <excludeFolder url="file://$MODULE_DIR$/examples/dist" />
      <excludeFolder url="file://$MODULE_DIR$/examples/.cache" />
      <excludeFolder url="file://$MODULE_DIR$/coverage" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>