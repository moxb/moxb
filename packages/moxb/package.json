{
  "name": "@moxb/moxb",
  "version": "0.2.0-beta.59",
  "description": "MobX data binding",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "main": "dist",
  "typings": "dist/",
  "files": [
    "dist/**/*"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/moxb/moxb.git"
  },
  "author": "moxb (moxb.org)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/moxb/moxb/issues"
  },
  "homepage": "https://github.com/moxb/moxb#readme",
  "keywords": [
    "mobx",
    "UI"
  ],
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "transform": {
      "^.+\\.(ts|tsx)$": "ts-jest"
    },
    "testEnvironment": "node",
    "testMatch": [
      "**/src/**/__tests__/*.(ts|tsx)"
    ],
    "modulePaths": [
      "<rootDir>/peer-dependencies/node_modules"
    ]
  },
  "peerDependencies": {
    "mobx": "^5.15.6",
    "tslib": "^2.0.1"
  },
  "devDependencies": {
    "@types/atob": "2.1.2",
    "@types/jest": "26.0.13",
    "@types/lodash.flatten": "4.4.6",
    "@types/urijs": "1.19.11",
    "mobx": "5.13.1",
    "tslib": "1.10.0"
  },
  "dependencies": {
    "btoa-lite": "1.0.0",
    "history": "5.0.0",
    "lodash.flatten": "4.4.0",
    "moment": "2.29.1",
    "urijs": "1.19.2"
  }
}
