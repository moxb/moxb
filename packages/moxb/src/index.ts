export * from './impl';
export * from './types';
export * from './routing';
