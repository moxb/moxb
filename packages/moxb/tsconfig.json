{
    "compilerOptions": {
        "module": "commonjs",
        "target": "ES5",
        "jsx": "react",
        "lib": ["dom","es6","es2015","es2017", "dom.iterable","scripthost"],
        "inlineSourceMap": true,
        "importHelpers":true,
        "experimentalDecorators": true,
        "noImplicitAny": true,
        "noImplicitThis": true,
        "strict": true,
        "outDir": "./dist",
        "declaration": true,
        "skipLibCheck": true,
        "strictNullChecks": true,
        "removeComments": false,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "allowUnreachableCode": false,
        "baseUrl": "./src",
        "paths": {
            "*": [
                "*",
                "../../../node_modules/*"
            ]
        }
    },
    "exclude": [
        "./node_modules",
        "./dist"
    ]
}
