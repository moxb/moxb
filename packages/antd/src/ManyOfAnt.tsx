import { idToDomId, ManyOf } from '@moxb/moxb';
import { Checkbox, Col, Row, Select, Switch } from 'antd';
import { CheckboxGroupProps } from 'antd/lib/checkbox';
import { SwitchChangeEventHandler } from 'antd/lib/switch';
import { toJS } from 'mobx';
import { observer } from 'mobx-react';
import * as React from 'react';
import { BindAntProps, labelWithHelp, parseProps } from './BindAnt';
import { BindFormItemAntProps, FormItemAnt, parsePropsForChild } from './FormItemAnt';
import { SelectProps } from 'antd/lib/select';
import { AllowedSwitchProps } from './BoolAnt';

@observer
export class ManyOfAnt extends React.Component<BindAntProps<ManyOf> & SelectProps<any>> {
    render() {
        const { operation, invisible, mode, children, defaultValue, reason, ...props } = parseProps(
            this.props,
            this.props.operation
        );
        if (invisible) {
            return null;
        }
        // make sure the value is not a mobx object...
        const value = toJS(operation.value);
        return (
            <span title={reason}>
                <Select
                    data-testid={operation.id}
                    onChange={(selectionValue: any) => operation.setValue(selectionValue)}
                    value={value || undefined}
                    placeholder={operation.placeholder}
                    defaultValue={defaultValue}
                    mode={mode}
                    {...props}
                >
                    {operation.choices.map((opt: any) => (
                        <Select.Option
                            data-testid={idToDomId(operation.id + '.' + opt.value)}
                            key={opt.value}
                            value={opt.value}
                        >
                            {opt.label}
                        </Select.Option>
                    ))}
                    {children}
                </Select>
            </span>
        );
    }
}

export interface ManyOfCheckboxAntProps {
    /**
     * Should we align the items vertically, instead of horizontally?
     */
    vertical?: boolean;
}

@observer
export class ManyOfCheckboxAnt extends React.Component<
    BindAntProps<ManyOf> & CheckboxGroupProps & ManyOfCheckboxAntProps
> {
    render() {
        const { operation, invisible, children, defaultValue, reason, vertical, ...props } = parseProps(
            this.props,
            this.props.operation
        );
        if (invisible || operation.invisible) {
            return null;
        }
        // make sure the value is not a mobx object...
        const value = toJS(operation.value);

        const choices = vertical
            ? operation.choices.map((opt: any) => (
                  <Row title={reason} key={opt.value} style={{ width: '100%' }}>
                      <Col span={24}>
                          <Checkbox data-testid={idToDomId(operation.id + '.' + opt.value)} value={opt.value}>
                              {labelWithHelp(opt.label, opt.help)}
                          </Checkbox>
                      </Col>
                  </Row>
              ))
            : operation.choices.map((opt: any) => (
                  <Col title={reason} key={opt.value}>
                      <Checkbox data-testid={idToDomId(operation.id + '.' + opt.value)} value={opt.value}>
                          {labelWithHelp(opt.label, opt.help)}
                      </Checkbox>
                  </Col>
              ));

        return (
            <Checkbox.Group
                data-testid={operation.id}
                onChange={(selectionValue: any) => operation.setValue(selectionValue)}
                value={value}
                defaultValue={defaultValue}
                {...props}
            >
                <Row>
                    {choices}
                    {children}
                </Row>
            </Checkbox.Group>
        );
    }
}

@observer
export class ManyOfFormAnt extends React.Component<SelectProps<ManyOf> & BindAntProps<ManyOf> & BindFormItemAntProps> {
    render() {
        const { operation, invisible, ...props } = parsePropsForChild(this.props, this.props.operation);
        if (invisible) {
            return null;
        }
        return (
            <FormItemAnt operation={operation} {...(this.props as any)}>
                <ManyOfAnt operation={operation} {...props} />
            </FormItemAnt>
        );
    }
}

@observer
export class ManyOfCheckboxFormAnt extends React.Component<
    CheckboxGroupProps & BindAntProps<ManyOf> & BindFormItemAntProps & ManyOfCheckboxAnt
> {
    render() {
        const { operation, invisible, ...props } = parsePropsForChild(this.props, this.props.operation);
        if (invisible) {
            return null;
        }
        return (
            <FormItemAnt operation={operation} {...(this.props as any)}>
                <ManyOfCheckboxAnt operation={operation} {...props} />
            </FormItemAnt>
        );
    }
}

export interface ManyOfSwitchAntProps {
    /**
     * Should we align the choices vertically, instead of horizontelly?
     */
    vertical?: boolean;
}

@observer
export class ManyOfSwitchAnt extends React.Component<BindAntProps<ManyOf> & AllowedSwitchProps & ManyOfSwitchAntProps> {
    private _toggle(value: string) {
        const { operation, readOnly, disabled } = parseProps(this.props, this.props.operation);
        if (readOnly || disabled) {
        } else {
            operation.toggle(value);
        }
    }

    private readonly _switchHandlerCache: Record<string, SwitchChangeEventHandler> = {};

    private _getSwitchHandler(value: string): SwitchChangeEventHandler {
        let result = this._switchHandlerCache[value];
        if (!result) {
            result = this._switchHandlerCache[value] = () => {
                this._toggle(value);
            };
        }
        return result;
    }

    private readonly _clickHandlers: Record<string, React.MouseEventHandler> = {};

    private _getClickHandler(value: string): React.MouseEventHandler {
        let result = this._clickHandlers[value];
        if (!result) {
            result = this._clickHandlers[value] = () => {
                this._toggle(value);
            };
        }
        return result;
    }

    render() {
        const { operation, invisible, disabled, children, vertical, reason, ...props } = parseProps(
            this.props,
            this.props.operation
        );
        if (invisible || operation.invisible) {
            return null;
        }
        const switchClassName = disabled ? 'ant-checkbox-disabled' : '';
        const labelClassName = disabled ? '' : 'ant-checkbox-wrapper';

        const choices = vertical
            ? operation.choices.map((opt: any) => (
                  <Row key={opt.value} style={{ width: '100%' }}>
                      <Col span={24}>
                          <span title={reason}>
                              <span className={switchClassName}>
                                  <Switch
                                      data-testid={idToDomId(operation.id + '.' + opt.value)}
                                      checked={operation.isSelected(opt.value)}
                                      onChange={this._getSwitchHandler(opt.value)}
                                      disabled={disabled}
                                      size={'small'}
                                      {...props}
                                  />
                              </span>
                              &nbsp; &nbsp;
                              <span onClick={this._getClickHandler(opt.value)} className={labelClassName}>
                                  {labelWithHelp(opt.label, opt.help)}
                              </span>
                          </span>
                      </Col>
                  </Row>
              ))
            : operation.choices.map((opt: any) => (
                  <Col key={opt.value}>
                      <span title={reason}>
                          <span className={switchClassName}>
                              <Switch
                                  data-testid={idToDomId(operation.id + '.' + opt.value)}
                                  checked={operation.isSelected(opt.value)}
                                  onChange={this._getSwitchHandler(opt.value)}
                                  disabled={disabled}
                                  size={'small'}
                                  {...props}
                              />
                          </span>
                          &nbsp; &nbsp;
                          <span onClick={this._getClickHandler(opt.value)} className={labelClassName}>
                              {labelWithHelp(opt.label, opt.help)}
                          </span>
                          &nbsp; &nbsp; &nbsp; &nbsp;
                      </span>
                  </Col>
              ));

        return (
            <Row>
                {choices}
                {children}
            </Row>
        );
    }
}

@observer
export class ManyOfSwitchFormAnt extends React.Component<
    AllowedSwitchProps & BindAntProps<ManyOf> & BindFormItemAntProps & ManyOfSwitchAntProps
> {
    render() {
        const { operation, invisible, ...props } = parsePropsForChild(this.props, this.props.operation);
        if (invisible) {
            return null;
        }
        return (
            <FormItemAnt operation={operation} {...(this.props as any)}>
                <ManyOfSwitchAnt operation={operation} {...props} />
            </FormItemAnt>
        );
    }
}
