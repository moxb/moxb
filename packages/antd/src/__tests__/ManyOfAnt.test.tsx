import { ManyOfImpl } from '@moxb/moxb';
import { shallow } from 'enzyme';
import * as React from 'react';
import { ManyOfAnt, ManyOfCheckboxAnt, ManyOfFormAnt } from '../ManyOfAnt';
import { shallowMoxbToJson } from './enzymeHelper';

const manyChoices: any[] = [];
const allChoices = [
    { label: 'Banana', value: 'b' },
    { label: 'Apples', value: 'a' },
    { label: 'Peaches', value: 'p' },
];

describe('ManyOfAnt', function () {
    it('should render a select control by default', function () {
        const operation = new ManyOfImpl({
            id: 'Id.manyOf',
            label: 'Choose your snack',
            choices: () => allChoices,
            initialValue: () => manyChoices,
            placeholder: () => 'Please select',
            onSave: () => {},
        });
        expect(shallowMoxbToJson(shallow(<ManyOfAnt operation={operation} />))).toMatchSnapshot();
    });

    it('should return null if invisible', function () {
        const operation = new ManyOfImpl({
            id: 'Id.manyOf',
            label: 'Choose your snack',
            choices: () => allChoices,
            initialValue: () => manyChoices,
            placeholder: () => 'Please select',
            onSave: () => {},
            invisible: () => true,
        });
        expect(shallow(<ManyOfAnt operation={operation} />).type()).toBeNull();
    });
});

describe('ManyOfCheckboxAnt', function () {
    it('should render a checkboxes by default', function () {
        const operation = new ManyOfImpl({
            id: 'Id.manyOf',
            label: 'Choose your snack',
            choices: () => allChoices,
            initialValue: () => manyChoices,
            placeholder: () => 'Please select',
            onSave: () => {},
        });
        expect(shallowMoxbToJson(shallow(<ManyOfCheckboxAnt operation={operation} />))).toMatchSnapshot();
    });

    it('should return null if invisible', function () {
        const operation = new ManyOfImpl({
            id: 'Id.manyOf',
            label: 'Choose your snack',
            choices: () => allChoices,
            initialValue: () => manyChoices,
            placeholder: () => 'Please select',
            onSave: () => {},
            invisible: () => true,
        });
        expect(shallow(<ManyOfCheckboxAnt operation={operation} />).type()).toBeNull();
    });
});

describe('ManyOfFormAnt', function () {
    it('should render a checkboxes by default', function () {
        const operation = new ManyOfImpl({
            id: 'Id.manyOf',
            label: 'Choose your snack',
            choices: () => allChoices,
            initialValue: () => manyChoices,
            placeholder: () => 'Please select',
            onSave: () => {},
        });
        expect(shallowMoxbToJson(shallow(<ManyOfFormAnt operation={operation} />))).toMatchSnapshot();
    });

    it('should return null if invisible', function () {
        const operation = new ManyOfImpl({
            id: 'Id.manyOf',
            label: 'Choose your snack',
            choices: () => allChoices,
            initialValue: () => manyChoices,
            placeholder: () => 'Please select',
            onSave: () => {},
            invisible: () => true,
        });
        expect(shallow(<ManyOfFormAnt operation={operation} />).type()).toBeNull();
    });
});
