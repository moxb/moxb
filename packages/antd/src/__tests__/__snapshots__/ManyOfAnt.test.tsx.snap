// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManyOfAnt should render a select control by default 1`] = `
<span>
  <ForwardRef(InternalSelect)
    data-testid="Id.manyOf"
    disabled={false}
    id="id-many_of"
    label="Choose your snack"
    onChange={[Function]}
    placeholder="Please select"
    readOnly={false}
    value={Array []}
  >
    <Option
      data-testid="id-many_of-b"
      key="b"
      value="b"
    >
      Banana
    </Option>
    <Option
      data-testid="id-many_of-a"
      key="a"
      value="a"
    >
      Apples
    </Option>
    <Option
      data-testid="id-many_of-p"
      key="p"
      value="p"
    >
      Peaches
    </Option>
  </ForwardRef(InternalSelect)>
</span>
`;

exports[`ManyOfCheckboxAnt should render a checkboxes by default 1`] = `
<Memo(ForwardRef(InternalCheckboxGroup))
  data-testid="Id.manyOf"
  disabled={false}
  id="id-many_of"
  label="Choose your snack"
  onChange={[Function]}
  readOnly={false}
  value={Array []}
>
  <Row>
    <Col
      key="b"
    >
      <Checkbox
        data-testid="id-many_of-b"
        value="b"
      >
        Banana
      </Checkbox>
    </Col>
    <Col
      key="a"
    >
      <Checkbox
        data-testid="id-many_of-a"
        value="a"
      >
        Apples
      </Checkbox>
    </Col>
    <Col
      key="p"
    >
      <Checkbox
        data-testid="id-many_of-p"
        value="p"
      >
        Peaches
      </Checkbox>
    </Col>
  </Row>
</Memo(ForwardRef(InternalCheckboxGroup))>
`;

exports[`ManyOfFormAnt should render a checkboxes by default 1`] = `
<Memo(observerComponent)
  operation="bound ManyOfImpl"
>
  <Memo(observerComponent)
    disabled={false}
    id="id-many_of"
    operation="bound ManyOfImpl"
    readOnly={false}
  />
</Memo(observerComponent)>
`;
